src/pass/lift_if_then_else.cc

```diff
@@ -0,0 +1,422 @@
+  Stmt HoistIf(const Stmt& if_stmt);
+
+  HoistMap if2for_map_;
+  HoistMap top_for_var_map_;
+  HoistMap for_tracking_map_;
+  HoistMap for2if_map_;
+  VarMap cond_var_map_;
+  std::vector<Stmt> ordered_for_list_;
+};
+
+// Check whether a given IfThenElse stmt is the first one appearing
+// in a For stmt.
+bool is_first_if(const Stmt& for_stmt, const Stmt& if_stmt) {
+  std::vector<const Node*> if_node_list;
+  const For* for_node = for_stmt.as<For>();
+  CHECK(for_node);
+  CHECK(if_stmt.as<IfThenElse>());
+
+  PostOrderVisit(for_node->body, [&](const NodeRef& node) {
+    if (node.as<IfThenElse>()) {
+      if_node_list.push_back(node.get());
+    }
+  });
+  return if_node_list.empty() ? false : if_stmt.get() == if_node_list.back();
+}
+
+// Update upper level For node when current For node is modified.
+// With this function we only need to visit and mutate top level For node
+// in the main VisitAndMutate function.
+Stmt update_for(const Stmt& parent_for_stmt, const Stmt& new_if_stmt) {
+  std::vector<const Node*> for_node_list;
+  const For* parent_for_node = parent_for_stmt.as<For>();
+  CHECK(parent_for_node);
+  CHECK(new_if_stmt.as<IfThenElse>());
+
+  PostOrderVisit(parent_for_node->body, [&](const NodeRef& node) {
+    if (node.as<For>()) {
+      for_node_list.push_back(node.get());
+    }
+  });
+
+  PackedFunc replace_target_for = PackedFunc(
+    [&](TVMArgs args, TVMRetValue *ret){
+      const NodeRef& current_for = args[0];
+      if (current_for.get() == for_node_list.back()) {
+        *ret = new_if_stmt;
+      }
+    });
+
+  return IRTransform(parent_for_stmt, nullptr, replace_target_for,
+                     {Expr("For")});
+}
+
+// Remove IfThenElse node from a For node.
+// A pair of For nodes will be generated.
+std::pair<Stmt, Stmt> RemoveIf(const Stmt& for_stmt, const Stmt& if_stmt) {
+  Stmt then_for;
+  Stmt else_for;
+  CHECK(if_stmt.as<IfThenElse>());
+
+  PackedFunc replace_then_case = PackedFunc(
+    [&](TVMArgs args, TVMRetValue *ret){
+      const NodeRef& node  = args[0];
+      if (node == if_stmt) {
+        *ret = node.as<IfThenElse>()->then_case;
+      }
+    });
+
+  PackedFunc replace_else_case = PackedFunc(
+    [&](TVMArgs args, TVMRetValue *ret){
+      const NodeRef& node  = args[0];
+      if (node == if_stmt) {
+        *ret = node.as<IfThenElse>()->else_case;
+      }
+    });
+
+  then_for = IRTransform(for_stmt, nullptr, replace_then_case,
+                         {Expr("IfThenElse")});
+  if (if_stmt.as<IfThenElse>()->else_case) {
+    else_for = IRTransform(for_stmt, nullptr, replace_else_case,
+                           {Expr("IfThenElse")});
+  }
+
+  return std::make_pair(then_for, else_for);
+}
+
+// Locate all For nodes and capture child IfThenElse nodes.
+void IfThenElseLHoist::SelectCandidates(const Stmt& stmt) {
+  PostOrderVisit(stmt, [&](const NodeRef& node){
+    const For* for_node = node.as<For>();
+    if (for_node) {
```

Comment:
Nit - How about we return if it is not a for loop. Might increase readability of the code. 

`if (for_node == nullptr) return;`